###Analytic Decision Making

Exercise 1: Operations Excellence

1. Write the optimization models for each plant, i.e., make sure to detail the decision
variables, the objective function and the constraints.
2. Provide the optimal solutions and profit levels for each factory. Give the values of the
decision variables and the optimal objective function value.
3. Now, write the optimization model for the joint optimization of both sites, and provide
the optimal solution and profit levels.


In [19]:
#Question 1 contains help from previous lecture example and minor help from chatgpt
import math
import numpy
import pandas as pd
from cvxopt import matrix
from cvxopt.modeling import variable
from cvxopt.modeling import op

#decision variables
standard_BIC = variable(1, 'Standard_BIC')
deluxe_BIC = variable(1, 'Deluxe_BIC')

# Definition of the Constraints

c1 = (4 * standard_BIC + 2 * deluxe_BIC <= 80)  
c2 = (2 * standard_BIC + 5 * deluxe_BIC <= 60)  
c3 = (4 * standard_BIC + 4 * deluxe_BIC <= 75)  
c4 = (standard_BIC >= 0)
c5 = (deluxe_BIC >= 0)

const_set = [c1, c2, c3, c4, c5]

# Definition of the Objective Function
obj_func = 10 * standard_BIC + 15 * deluxe_BIC

# Solving the problem
problem = op(-obj_func, const_set) 
problem.solve()

OA=op(-obj_func,const_set) 
OA.solve()
OA.status
print(OA.objective.value())

print("BIC Optimal Production Plan:")
print("BIC Standard units:", standard_BIC.value[0])
print("BIC Deluxe units:", deluxe_BIC.value[0])
print("BIC Maximized Profit:", -problem_BIC.objective.value()[0])

     pcost       dcost       gap    pres   dres   k/t
 0: -2.3150e+02 -5.3116e+02  1e+02  1e-01  1e+00  1e+00
 1: -2.2424e+02 -2.4880e+02  7e+00  1e-02  9e-02  4e-01
 2: -2.2499e+02 -2.2530e+02  8e-02  1e-04  1e-03  6e-03
 3: -2.2500e+02 -2.2500e+02  8e-04  1e-06  1e-05  6e-05
 4: -2.2500e+02 -2.2500e+02  8e-06  1e-08  1e-07  6e-07
 5: -2.2500e+02 -2.2500e+02  8e-08  1e-10  1e-09  6e-09
Optimal solution found.
     pcost       dcost       gap    pres   dres   k/t
 0: -2.3150e+02 -5.3116e+02  1e+02  1e-01  1e+00  1e+00
 1: -2.2424e+02 -2.4880e+02  7e+00  1e-02  9e-02  4e-01
 2: -2.2499e+02 -2.2530e+02  8e-02  1e-04  1e-03  6e-03
 3: -2.2500e+02 -2.2500e+02  8e-04  1e-06  1e-05  6e-05
 4: -2.2500e+02 -2.2500e+02  8e-06  1e-08  1e-07  6e-07
 5: -2.2500e+02 -2.2500e+02  8e-08  1e-10  1e-09  6e-09
Optimal solution found.
[-2.25e+02]

BIC Optimal Production Plan:
BIC Standard units: 11.250000001823
BIC Deluxe units: 7.4999999978449425
BIC Maximized Profit: 224.99999998046462


In [18]:
#decision variables
standard_CIB = variable(1, 'Standard_CIB')
deluxe_CIB = variable(1, 'Deluxe_CIB')

# Definition of the Constraints
c1 = (5 * standard_CIB + 3 * deluxe_CIB <= 60)  
c2 = (5 * standard_CIB + 6 * deluxe_CIB <= 75)  
c3 = (4 * standard_CIB + 4 * deluxe_CIB <= 45)  
c4 = (standard_CIB >= 0)
c5 = (deluxe_CIB >= 0)

const_set = [c1, c2, c3, c4, c5]

# Definition of the Objective Function
obj_func = 10 * standard_CIB + 15 * deluxe_CIB

# Solving the problem
problem = op(-obj_func, const_set) 
problem.solve()

OA=op(-obj_func,const_set) 
OA.solve()
OA.status
print(OA.objective.value())

print("CIB Optimal Production Plan:")
print("CIB Standard units:", standard_CIB.value[0])
print("CIB Deluxe units:", deluxe_CIB.value[0])
print("CIB Maximized Profit:", -problem.objective.value()[0])

     pcost       dcost       gap    pres   dres   k/t
 0: -1.5641e+02 -5.0061e+02  1e+02  2e-01  2e+00  1e+00
 1: -1.7716e+02 -2.3795e+02  3e+01  3e-02  3e-01  2e+00
 2: -1.6813e+02 -1.8563e+02  8e+00  1e-02  1e-01  1e+00
 3: -1.6875e+02 -1.6893e+02  8e-02  1e-04  1e-03  1e-02
 4: -1.6875e+02 -1.6875e+02  8e-04  1e-06  1e-05  1e-04
 5: -1.6875e+02 -1.6875e+02  8e-06  1e-08  1e-07  1e-06
 6: -1.6875e+02 -1.6875e+02  8e-08  1e-10  1e-09  1e-08
Optimal solution found.
     pcost       dcost       gap    pres   dres   k/t
 0: -1.5641e+02 -5.0061e+02  1e+02  2e-01  2e+00  1e+00
 1: -1.7716e+02 -2.3795e+02  3e+01  3e-02  3e-01  2e+00
 2: -1.6813e+02 -1.8563e+02  8e+00  1e-02  1e-01  1e+00
 3: -1.6875e+02 -1.6893e+02  8e-02  1e-04  1e-03  1e-02
 4: -1.6875e+02 -1.6875e+02  8e-04  1e-06  1e-05  1e-04
 5: -1.6875e+02 -1.6875e+02  8e-06  1e-08  1e-07  1e-06
 6: -1.6875e+02 -1.6875e+02  8e-08  1e-10  1e-09  1e-08
Optimal solution found.
[-1.69e+02]

CIB Standard units: -2.4756703304694626e-09
CIB

In [20]:


#decision variables
standard_BIC = variable(1, 'Standard_BIC')
deluxe_BIC = variable(1, 'Deluxe_BIC')
standard_CIB = variable(1, 'Standard_CIB')
deluxe_CIB = variable(1, 'Deluxe_CIB')

# Definition of the Constraints

c1 = (4 * standard_BIC + 2 * deluxe_BIC <= 80)  
c2 = (2 * standard_BIC + 5 * deluxe_BIC <= 60)  
c3 = (4 * standard_BIC + 4 * deluxe_BIC <= 75) 

c4 = (5 * standard_CIB + 3 * deluxe_CIB <= 60)  
c5 = (5 * standard_CIB + 6 * deluxe_CIB <= 75)  
c6 = (4 * standard_CIB + 4 * deluxe_CIB <= 45)  

c7 = (standard_BIC >= 0)
c8 = (deluxe_BIC >= 0)
c9 = (standard_CIB >= 0)
c10 = (deluxe_CIB >= 0)

const_set = [c1, c2, c3, c4, c5, c6, c7, c8, c9, c10]

# Definition of the Objective Function
obj_func = 10 * standard_BIC + 15 * deluxe_BIC + 10 * standard_CIB + 15 * deluxe_CIB

# Solving the problem
problem = op(-obj_func, const_set)  # Use negative profit for maximization in cvxopt
problem.solve()

#Solving the problem
OA=op(-obj_func,const_set) 
OA.solve()
OA.status
print(OA.objective.value())


print("BIC Standard units:", standard_BIC.value[0])
print("BIC Deluxe units:", deluxe_BIC.value[0])
print("CIB Standard units:", standard_CIB.value[0])
print("CIB Deluxe units:", deluxe_CIB.value[0])
print("Maximized Profit:", -problem.objective.value()[0])

     pcost       dcost       gap    pres   dres   k/t
 0: -3.8791e+02 -1.1432e+03  3e+02  2e-01  2e+00  1e+00
 1: -4.0846e+02 -6.6129e+02  9e+01  6e-02  6e-01  3e+00
 2: -3.9606e+02 -4.9296e+02  4e+01  2e-02  2e-01  2e+00
 3: -3.9378e+02 -3.9542e+02  5e-01  4e-04  4e-03  6e-02
 4: -3.9375e+02 -3.9377e+02  5e-03  4e-06  4e-05  6e-04
 5: -3.9375e+02 -3.9375e+02  5e-05  4e-08  4e-07  6e-06
 6: -3.9375e+02 -3.9375e+02  5e-07  4e-10  4e-09  6e-08
Optimal solution found.
     pcost       dcost       gap    pres   dres   k/t
 0: -3.8791e+02 -1.1432e+03  3e+02  2e-01  2e+00  1e+00
 1: -4.0846e+02 -6.6129e+02  9e+01  6e-02  6e-01  3e+00
 2: -3.9606e+02 -4.9296e+02  4e+01  2e-02  2e-01  2e+00
 3: -3.9378e+02 -3.9542e+02  5e-01  4e-04  4e-03  6e-02
 4: -3.9375e+02 -3.9377e+02  5e-03  4e-06  4e-05  6e-04
 5: -3.9375e+02 -3.9375e+02  5e-05  4e-08  4e-07  6e-06
 6: -3.9375e+02 -3.9375e+02  5e-07  4e-10  4e-09  6e-08
Optimal solution found.
[-3.94e+02]

BIC Standard units: 11.249999996966041
BIC Delu

4. Please comment on the results? How do they compare to the results obtained in Question 2


From the results, The individual optimiztion results are 
BIC Site Profit: 224
CIB Site Profit: 168
Combined Profit: 392

The joint optimization results is 393.75 in maximized profit.  

The joint optimiztion approach results in a slightly higher profit. This may suggest some minor addition effiency when optimizing both CIB and BIC together. We can see that CIB site should not produce standard units, and focus on deluxe unit. The join optimization can improve resource allocation. 


5. In one paragraph (no more than 300 words), detail the recommendations you would provide to the CEO of the company to improve the firm’s operations? 

Based on the joint optimization results, I would recommend the CEO of the compnay to implement the joint optimization. Eventhough it is only a slight improve, but it can acheive a marginal increase in overall profit. Next, the CEO should start to apply centralized planning to optimize resource allocation across all sites. The results show that current allocaiton may lead to slight ineffiency, and the best way is to let the production sites focus on the most profitable mix of products. In this case, the CEO should review and adjust allocation of raw materials, cooking, and filtering capacities regularly. Is it also worth to inspect for the reason for the optimization suggesting CIB site to not product standard unit. Finally, the CEO should continue monitor the production effiency and resource allocation in the future to ensure good profit. 










Exercise 2: Hotel L’Escargot 

Profitability = 39.05 - (5.41 xState Population per Inn)
+ (5.86 × Price of the Inn)
− (3.09 × Square Root of the Median Income in the area)
+ (1.75 × College Students in the Area)

All independent variables are significant and were normalized to have mean zero and
standard deviation 1.
1. According to the regression equation given above, which variable positively affect Profitability? Which
variable negatively affect Profitability? Does this intuitively make sense? Why?

According to the regression equation given above, Price of the Inn and College Students in the area have the positively affect, while Square root of the median income in the area. 


2. Using this regression equation, L’Escargot created a spreadsheet model to predict profitability.
L’Escargot collected data for several locations in California, which is provided in the excel
spreadsheet on Canvas “L’Escargot”. Using this spreadsheet, compute the profitability for each hotel.
Which one has the highest profitability? Which one has the lowest profitability?

In [12]:
#The whole question 2 contains partial help from chatgpt and claude 3

import os
import pandas as pd
import numpy as np
import cvxpy as cp

directory_path = '/Users/richa/Downloads' 
os.chdir(directory_path)
current_directory = os.getcwd()
print("Current working directory:", current_directory)

lescargot = 'Lescargot.xlsx'
df = pd.read_excel(lescargot, header=2)

Current working directory: c:\Users\richa\Downloads


C:\Users\richa\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.12_qbz5n2kfra8p0\LocalCache\local-packages\Python312\site-packages\openpyxl\worksheet\_reader.py:329: UserWarning: Unknown extension is not supported and will be removed
  warn(msg)


In [14]:
#  profitability
def compute_profitability(row):
    state_pop_per_inn = row['State Population Per Inn (normalized)']
    price = row['Price (normalized)']
    sqrt_median_income = row['Square Root of Median Income (normalized)']
    college_students = row['College Students in Area (normalized)']

    profitability = 39.05 - (5.41 * state_pop_per_inn) + (5.86 * price) - (3.09 * sqrt_median_income) + (1.75 * college_students)
    return profitability

# Apply the function to each row of the DataFrame
df['Profitability'] = df.apply(compute_profitability, axis=1)

# Sort the DataFrame by profitability
sorted_data = df.sort_values(by='Profitability', ascending=False)

# Print the most and least profitable hotels
most_profitable = sorted_data.iloc[0]
least_profitable = sorted_data.iloc[-1]

print("\nMost Profitable Hotel:")
print(most_profitable)
print("\nLeast Profitable Hotel:")
print(least_profitable)




Most Profitable Hotel:
Hotel                                                         2
Location                                     Fresno, California
Price                                                  10000000
Price (normalized)                                     1.699076
Square Root of Median Income (normalized)             -0.408199
College Students in Area (normalized)                  0.311669
State Population Per Inn (normalized)                 -0.474279
Profitability                                         53.379192
Name: 1, dtype: object

Least Profitable Hotel:
Hotel                                                              8
Location                                     Los Angeles, California
Price                                                        1750000
Price (normalized)                                         -0.634128
Square Root of Median Income (normalized)                   0.167686
College Students in Area (normalized)                       3.106215
St

3. L’Escargot has a budget of $10 million to spend on hotels. Suppose we use a “greedy” approach
where we select the most profitable hotels until we ran out of budget. So we would start by the most
profitable, and then if we had enough budget left, we would buy the hotel we predict to be the second
most profitable, and so on.
a. Describe what we would do with this approach, i.e., which hotels would we purchase?
b. What would our total predicted profitability be? (This is the sum of the predicted
profitability of all hotels we purchase.)
c. If we are trying to maximize our total predicted profitability, is this a good approach? How
about if we were trying to maximize the average predicted profitability of the hotels we
select? How about if we had a budget of $20 million instead of $10 million?



a. By doing this we can only afford hotel 2

b. The profitability result is 53.3

c. The result suggests that maximizing the total prediicted profitability, avg profitability or turn budget to 20 million do not provide best approach. 

In [17]:

# Define decision variables
number_hotel = len(df)
x = cp.Variable(number_hotel, boolean=True)  

# Objective function
profitability = df['Profitability'].values
obj_func = cp.Maximize(profitability @ x)

# Budget constraint
prices = df['Price'].values
budget = 10000000  
budget_constraint = cp.sum(prices * x) <= budget

# Problem formulation
problem = cp.Problem(obj_func, [budget_constraint])

# Solve the problem
problem.solve()

# Print the selected hotels and their total predicted profitability
selected_hotels = data.iloc[x.value.astype(bool)]
total_predicted_profitability = sum(selected_hotels['Profitability'])

print("Decision variable values (x):")
print(x.value)
print("\nSelected hotels:")
print(selected_hotels[['Hotel', 'Price', 'Profitability']])
print(f"\nTotal price: ${sum(selected_hotels['Price']):,.2f}")
print(f"Total predicted profitability: {total_predicted_profitability:.2f}")


Decision variable values (x):
[0. 0. 0. 0. 1. 0. 0. 0. 0. 1. 1. 1. 1. 0. 1. 1.]

Selected hotels:
    Hotel    Price  Profitability
4       5   325000      37.344988
9      10  1650000      38.880673
10     11  1125000      38.010600
11     12  2500000      40.289363
12     13  1975000      39.419290
14     15  1475000      38.590649
15     16   750000      37.389119

Total price: $9,800,000.00
Total predicted profitability: 269.92


C:\Users\richa\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.12_qbz5n2kfra8p0\LocalCache\local-packages\Python312\site-packages\cvxpy\expressions\expression.py:650: UserWarning: 
This use of ``*`` has resulted in matrix multiplication.
Using ``*`` for matrix multiplication has been deprecated since CVXPY 1.1.
    Use ``*`` for matrix-scalar and vector-scalar multiplication.
    Use ``@`` for matrix-matrix and matrix-vector multiplication.
    Use ``multiply`` for elementwise multiplication.
This code path has been hit 3 times so far.

  warnings.warn(msg, UserWarning)


The optimal solution is by selecting hotel 5, 10, 11, 12, 13, 15, 16. And the predicted profitability adds up to around 269 wish price of around 98 million.

5. L’Escargot thinks that buying too many hotels in one city is probably not a good idea and would
prefer to diversify across as many cities as possible. Add constraint(s) to your model to limit the
number of hotels purchased in any city to at most 2.
a. What are the constraints that you need to add to the model? Intuitively, do you expect the
new optimal objective function value to be larger, smaller or the same as before?
b. Write the new optimization model.
c. Solve the new model. Give the values of the decision variables and the optimal objective
function value. How does this compare to the previous solution?

a.     city_constraints.append(cp.sum(x[city_indices]) <= 2)
I expect it to be smaller. 


In [18]:
# Define decision variables
number_hotel = len(df)
x = cp.Variable(number_hotel, boolean=True)  

# Objective function
profitability = df['Profitability'].values
objective = cp.Maximize(profitability @ x)

# Budget constraint
prices = df['Price'].values
budget = 10000000
budget_constraint = cp.sum(prices * x) <= budget

# City constraint: At most 2 hotels per city
cities = df['Location'].unique()
city_constraints = []
for city in cities:
    city_indices = df[df['Location'] == city].index
    city_constraints.append(cp.sum(x[city_indices]) <= 2)

# Problem formulation
constraints = [budget_constraint] + city_constraints
problem = cp.Problem(objective, constraints)

# Solve the problem
problem.solve()

# Print the selected hotels and their total predicted profitability
selected_hotels = df.iloc[x.value.astype(bool)]
total_predicted_profitability = sum(selected_hotels['Profitability'])

print("Decision variable values (x):")
print(x.value)
print("\nSelected hotels:")
print(selected_hotels[['Hotel', 'Price', 'Profitability', 'Location']])
print(f"\nTotal price: ${sum(selected_hotels['Price']):,.2f}")
print(f"Total predicted profitability: {total_predicted_profitability:.2f}")

Decision variable values (x):
[1. 0. 0. 0. 1. 0. 1. 1. 0. 1. 1. 0. 0. 0. 0. 0.]

Selected hotels:
    Hotel    Price  Profitability                      Location
0       1  2925000      44.242369            Eureka, California
4       5   325000      37.344988            Fresno, California
6       7  1950000      23.776866       Los Angeles, California
7       8  1750000      23.445409       Los Angeles, California
9      10  1650000      38.880673  South Lake Tahoe, California
10     11  1125000      38.010600  South Lake Tahoe, California

Total price: $9,725,000.00
Total predicted profitability: 205.70


C:\Users\richa\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.12_qbz5n2kfra8p0\LocalCache\local-packages\Python312\site-packages\cvxpy\expressions\expression.py:650: UserWarning: 
This use of ``*`` has resulted in matrix multiplication.
Using ``*`` for matrix multiplication has been deprecated since CVXPY 1.1.
    Use ``*`` for matrix-scalar and vector-scalar multiplication.
    Use ``@`` for matrix-matrix and matrix-vector multiplication.
    Use ``multiply`` for elementwise multiplication.
This code path has been hit 4 times so far.

  warnings.warn(msg, UserWarning)


The final optimal function result is 205.70, which is reduced alot from question 3. This shows that limit the number of hotels to 2 per city does decrease the overall profitability. This reflects that diversified investement strategy is more prefered. 

Exercise 3: Matching for a Dating Platform

1. Assume all of the individuals in the data set would like to be matched to people of
the opposite gender. What matches would you choose to maximize the aggregate
match score?       

In [121]:
#Whole question 3 contains help from lecture example code by Jas. 

import cvxpy as cvx

# Load data
data_url = 'https://raw.githubusercontent.com/ormarketing/OD/master/OD.csv'
df = pd.read_csv(data_url)

# Separate males and females
females = df[df['Gender'] == 'Female']
males = df[df['Gender'] == 'Male']

# Initialize variables
num_females = len(females)
num_males = len(males)
activities = ['Sports', 'Theater', 'Religion', 'Outdoors', 'Eating Out']

# Create compatibility score matrix
compatibility_matrix = np.zeros((num_females, num_males))

for i in range(num_females):
    for j in range(num_males):
        # Calculate the compatibility score
        compatibility_matrix[i, j] = sum(females.iloc[i][activities] * males.iloc[j][activities])

# Optimization variables
X = cvx.Variable((num_females, num_males), boolean=True)

# Objective: Maximize sum of compatibility scores
objective = cvx.Maximize(cvx.sum(cvx.multiply(X, compatibility_matrix)))

# Constraints
# Each female is matched exactly once
constraints = [cvx.sum(X, axis=1) == 1]

# Each male is matched at most once
constraints += [cvx.sum(X, axis=0) <= 1]

# Solve the problem
problem = cvx.Problem(objective, constraints)
problem.solve()

# Output results
print("The maximum aggregate match score is:", problem.value)
matched_pairs = np.argwhere(X.value > 0.9)
for index in matched_pairs:
    female_name = females.iloc[index[0]]['Unnamed: 0']
    male_name = males.iloc[index[1]]['Unnamed: 0']
    print(f"Match {female_name} with {male_name}")

The maximum aggregate match score is: 48.0
Match Laura with Carl
Match Jennifer with Adam
Match Emma with Daniel
Match Olivia with Peter
Match Martha with Eric
Match Isabella with Ralph
Match Mary with James
Match Sophia with Robert


2. How would the results change if each person is presented with two possible
matches?

In [122]:
# Separate males and females
constraints = [cvx.sum(X, axis=1) == 2]
constraints += [cvx.sum(X, axis=0) <= 2]

# Solve the problem
problem = cvx.Problem(objective, constraints)
problem.solve()

# Output results
print("The maximum aggregate match score with two matches per person is:", problem.value)
matched_pairs = np.argwhere(X.value > 0.9)
for index in matched_pairs:
    female_name = females.iloc[index[0]]['Unnamed: 0']
    male_name = males.iloc[index[1]]['Unnamed: 0']
    print(f"Match {female_name} with {male_name}")


The maximum aggregate match score with two matches per person is: 82.0
Match Laura with Adam
Match Laura with Carl
Match Jennifer with Robert
Match Jennifer with Adam
Match Emma with James
Match Emma with Daniel
Match Olivia with Carl
Match Olivia with Ralph
Match Martha with Eric
Match Martha with Daniel
Match Isabella with Peter
Match Isabella with Ralph
Match Mary with James
Match Mary with Peter
Match Sophia with Robert
Match Sophia with Eric
